### ISSUE TITLE
java.lang.OutOfMemoryError: Java heap space

### ISSUE THREAD

**Issue by olsh** [gretard/sonar-tsql-plugin#19]
First of all, thank you for the great plugin!

We have a pretty large database project, and we get many warnings on analysis:
```
WARN: Unexpected error parsing file C:\path_to_project\file.sql
java.lang.OutOfMemoryError: Java heap space
```
and 
```
[11:29:52][Step 8/8] WARNING: WARN: [JOURNAL_FLUSHER] WARNING Journal flush operation took 6,295ms last 8 cycles average is 786ms
[11:31:05][Step 8/8] java.lang.OutOfMemoryError: GC overhead limit exceeded
[11:31:05][Step 8/8] WARNING: WARN: Unexpected error parsing file C:\path_to_project\file.sql
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.PredictionContext.mergeArrays(PredictionContext.java:376)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.PredictionContext.merge(PredictionContext.java:174)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.PredictionContext.mergeSingletons(PredictionContext.java:225)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.PredictionContext.merge(PredictionContext.java:155)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ATNConfigSet.add(ATNConfigSet.java:155)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1529)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closureCheckingStopState(ParserATNSimulator.java:1513)
[11:31:05][Step 8/8] 	at org.antlr.v4.runtime.atn.ParserATNSimulator.closure_(ParserATNSimulator.java:1586)
```

And at the end of the analysis, we get the error and complete analysis fails:

```
[12:39:03][Step 8/8] INFO: ------------------------------------------------------------------------
[12:39:03][Step 8/8] INFO: EXECUTION FAILURE
[12:39:03][Step 8/8] INFO: ------------------------------------------------------------------------
[12:39:03][Step 8/8] INFO: Total time: 1:13:32.583s
[12:39:05][Step 8/8] INFO: Final Memory: 757M/854M
[12:39:05][Step 8/8] INFO: ------------------------------------------------------------------------
[12:39:05][Step 8/8] ERROR: Error during SonarQube Scanner execution
[12:39:05][Step 8/8] ERROR: Java heap space
[12:39:05][Step 8/8] ERROR: 
[12:39:05][Step 8/8] The SonarQube Scanner did not complete successfully
[12:39:05][Step 8/8] 05:39:05.652  Creating a summary markdown file...
[12:39:05][Step 8/8] 05:39:05.658  Post-processing failed. Exit code: 1
[12:39:05][Step 8/8] Process exited with code 1
[12:39:05][Step 8/8] Process exited with code 1 (Step: SonarScanner for MSBuild: finish analysis)
[12:39:05][Step 8/8] Out of memory: Java heap space
[12:39:05][Step 8/8] Step SonarScanner for MSBuild: finish analysis failed
```

I'm not sure, but it seems like antlr parser is failed, so probably we can disable the feature somehow?

Thanks!

**Comment by gretard** [comment]
Hi,

I would suggest either increasing memory with -Xmx options for the sonar
scanner or skipping file with sonar.exclusions option (
https://docs.sonarqube.org/display/SONAR/Narrowing%2Bthe%2BFocus). Right
now there is no option to skip this part. I will include ability to skip
each sensor separately in the next release.

Best regards

**Comment by olsh** [comment]
Hi @gretard,

> I would suggest either increasing memory with -Xmx options for the sonar
scanner

I've tried increase memory for Java process for [`Compute Engine`](https://docs.sonarqube.org/display/SONAR/Java+Process+Memory), but this didn't help.

> or skipping file with sonar.exclusions option

There are many files in the log, and it seems like the file list is not deterministic.

> I will include ability to skip
each sensor separately in the next release.

Great, thank you!

**Comment by olsh** [comment]
Hi @gretard 

Thank you for your work!

`Max file size for analysis` didn't work for me (I've tried to set it to 5), because we don't have large files. Instead, we have many small files, and some projects contain 3000+ files.

With `Skip custom reporting` enabled the analysis pass without errors, but I it seems like the option disables rules by Microsoft, I don't see any issues in SonarQube at all. And I wonder, the option also disables the [Plugin rules](https://github.com/gretard/sonar-tsql-plugin/wiki/Plugin-rules)?

**Comment by gretard** [comment]
Hi, I would suggest increasing memory for the sonar scanner, not the Compute engine, by specifying -Xmx options (i.e. -Xmx2048m), please take a look at troubleshooting section at https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner. Or did you try that?

**Comment by olsh** [comment]
Hi @gretard 

> by specifying -Xmx options (i.e. -Xmx2048m), please take a look at troubleshooting section at https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner. Or did you try that?

This solved the issue.

> but I it seems like the option disables rules by Microsoft

I just forgot to set `/p:RunSqlCodeAnalysis=true`, everything works fine.

Thank you!
